### PR TITLE
Toggle messaging logs at runtime

### PR DESCRIPTION
**Pull request by fregante**
Blocked by https://github.com/pixiebrix/webext-messenger/pull/153

Until then, this won't run locally.

## What does this PR do?

Brings this over from the messenger:

- https://github.com/pixiebrix/webext-messenger/pull/134
- Closes https://github.com/pixiebrix/pixiebrix-extension/issues/6535

## Discussion

- If we're not going to add this to the main options, I can have it toggled via console like `window.pbToggleDark()`
- I thought about implementing this via `localStorage` so that it's synchronous, but `localStorage` is per-origin, so you'd have to enable it manually for every website

## Demo

https://github.com/pixiebrix/pixiebrix-extension/assets/1402241/30fa16bd-1465-4ea4-a670-20b8fe0c96a4

## Team Coordination

- [x] This PR requires a feature flag

## Checklist

- [x] New files added to `src/tsconfig.strictNullChecks.json` (if possible)
- [x] Designate a primary reviewer: @grahamlangford

### CODE REVIEW

File: src/extensionConsole/pages/settings/LoggingSettings.tsx [hunk 1]
```diff
@@ -69,10 +72,30 @@ const LoggingSettings: React.FunctionComponent = () => {
...
+              <BootstrapSwitchButton
+                size="sm"
+                onstyle="info"
+                offstyle="light"
+                onlabel=" "
+                offlabel=" "
+                checked={logMessenger}
+                onChange={async (toggleValue) => {
+                  setLogMessenger(toggleValue);
+                }}
```
Comment (fregante) [inline src/extensionConsole/pages/settings/LoggingSettings.tsx:98]: Does it make sense to add this toggle in this section? The description mentions "stored logs" whereas this feature refers to "browser console log".

The section is titled "Developer Settings" though

File: src/hooks/logging.ts [hunk 2]
```diff
@@ -15,39 +15,9 @@
...
-
-  const update = useCallback(
-    async (newConfig: LoggingConfig) => {
-      await setLoggingConfig(newConfig);
-      setConfig(newConfig);
-    },
-    [setConfig]
-  );
-
-  return [config, update];
```
Comment (fregante) [inline src/hooks/logging.ts:52]: I took all this logic and extracted it into a `useUpdatableAsyncState` hook. It feels like this pattern should be super common but I don't see any existing hooks using it.

- `useState`: you have to manually set any async values via `useAsyncEffect`
- `useAsyncState`: it doesn't take a setter
- `useAsyncExternalStore`: it doesn't take a setter


Usage:

```js
const [value, setValue] = useUpdatableAsyncState(asyncGetter, asyncSetter);
```

It doesn't expose any "in progress" state though.

File: src/hooks/useUpdatableAsyncState.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  const update = useCallback(
+    async (newValue: S) => {
+      await setter(newValue);
+      setValue(newValue);
+    },
+    [setValue, setter]
+  );
+
+  return [value, update];
+}
```
Comment (fregante) [inline src/hooks/useUpdatableAsyncState.ts:51]: Copy-pasted from `src/hooks/logging.ts` and made generic

File: src/background/background.ts [hunk 4]
```diff
@@ -45,6 +46,7 @@ import { initLogSweep } from "@/telemetry/logging";
 import { initModUpdater } from "@/background/modUpdater";
 
 void initLocator();
+void initMessengerLogging();
```
Comment (fregante) [inline src/background/background.ts:49]: Every context now has to check the storage and optionally turn logging on. This introduces a race condition.

Alternatives:

- await this call and delay every content (😬) 
- enable logging by default, disable it after checking the storage

The latter will leak some early logs, but it's probably better than losing them. Should I flip it?

File: src/utils/storageUtils.ts [hunk 5]
```diff
@@ -63,24 +63,33 @@ export type StorageKey = ManualStorageKey | ReduxStorageKey;
...
-  defaultValue?: T,
+  defaultValue: T,
+  area?: "local" | "session"
+): Promise<T>;
+async function readStorage(
+  storageKey: StorageKey,
+  defaultValue?: unknown,
   area: "local" | "session" = "local"
-): Promise<T | undefined> {
+): Promise<unknown> {
```
Comment (fregante) [inline src/utils/storageUtils.ts:77]: I had to update these types after adding my new files to the strict config.

Changes:

- don't return `T | undefined` when a `defaultValue` is provided; only return `T`

**Review by fregante — COMMENTED** [review]
This PR depends on an unpublished version of webext-messenger. The demo was recorded with a local installation.

Please review the setup and code and I'll update this PR when possible.

File: src/background/background.ts (hunk 4 above)
Comment (grahamlangford) [reply src/background/background.ts:49]: I'm inclined to agree to flipping it. @twschiller can you weigh in?

File: src/extensionConsole/pages/settings/LoggingSettings.tsx (hunk 1 above)
Comment (grahamlangford) [reply src/extensionConsole/pages/settings/LoggingSettings.tsx:98]: I think it makes more sense to include it here than in another existing card or creating a new card. We should provide a separate description for this toggle to differentiate.

File: src/hooks/useUpdatableAsyncState.ts (hunk 3 above)
Comment (grahamlangford) [reply src/hooks/useUpdatableAsyncState.ts:51]: I like this hook. We could standardize it with our other async state hooks by returning `[AsyncState<S | undefined>, Dispatch<SetStateAction<S | undefined>>]`.

File: src/background/background.ts (hunk 4 above)
Comment (grahamlangford) [reply src/background/background.ts:49]: I've spoken with @twschiller. The performance issues we're seeing are after startup, so it's ok to drop a few logs in the beginning. We should leave it toggled off by default.

File: src/hooks/useUpdatableAsyncState.ts (hunk 3 above)
Comment (fregante) [reply src/hooks/useUpdatableAsyncState.ts:51]: I'm not sure it's possible to provide [all of that data](https://github.com/pixiebrix/pixiebrix-extension/blob/005dccdce0d9fd14dc589731df120c896247dd91/src/types/sliceTypes.ts#L23) given two simple get/set functions. 

While working on strictNullChecks I found other places where this pattern is used so it's probably best to merge this as is and expand it later if needed/possible. I could look into it in a future PR.

**Review by grahamlangford — APPROVED** [review]

File: src/development/messengerLogging.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  if (await getMessengerLogging()) {
+    toggleLogging(true);
+  }
+
+  browser.storage.onChanged.addListener(async (changes) => {
+    if (changes[MESSENGER_LOGGING_KEY]) {
+      // eslint-disable-next-line security/detect-object-injection -- Constant key
+      toggleLogging(changes[MESSENGER_LOGGING_KEY].newValue);
+    }
+  });
```
Comment (fregante) [inline src/development/messengerLogging.ts:45]: This immediately toggles the logging without having to reload the context or extension

File: src/extensionConsole/pages/settings/LoggingSettings.tsx [hunk 7]
```diff
@@ -55,34 +56,53 @@ const LoggingSettings: React.FunctionComponent = () => {
...
+              size="sm"
+              onstyle="info"
+              offstyle="light"
+              onlabel=" "
+              offlabel=" "
+              checked={logMessenger}
+              onChange={setLogMessenger}
+            />
+          </Form.Group>
+        </Form>
```
Comment (fregante) [inline src/extensionConsole/pages/settings/LoggingSettings.tsx:101]: How does it look?

Screenshot 1

**Review by fregante — COMMENTED** [review]
Done. Merge at all

File: src/extensionConsole/pages/settings/LoggingSettings.tsx (hunk 7 above)
Comment (grahamlangford) [reply src/extensionConsole/pages/settings/LoggingSettings.tsx:101]: LGTM. See linting errors.

**Pull request merged**